src/many-identity/src/identity.rs

```diff
@@ -1,12 +1,153 @@
 //! An Identity is a signer that also has an address on the MANY protocol.
 use crate::Address;
+use coset::{CoseKey, CoseSign1};
 use many_error::ManyError;
 
-pub(crate) mod cose;
+pub trait Identity: Send {
+    fn address(&self) -> Address;
+    fn public_key(&self) -> Option<CoseKey>;
+    fn sign_1(&self, envelope: CoseSign1) -> Result<CoseSign1, ManyError>;
+}
 
-pub use cose::*;
+pub trait Verifier: Send {
+    fn sign_1(&self, envelope: &CoseSign1) -> Result<(), ManyError>;
+}
 
-pub trait Identity {
-    fn address(&self) -> Address;
-    fn sign(&self, _envelope: coset::CoseSign) -> Result<coset::CoseSign, ManyError>;
+#[derive(Debug, Clone)]
+pub struct AnonymousIdentity;
+
+impl Identity for AnonymousIdentity {
+    fn address(&self) -> Address {
+        Address::anonymous()
+    }
+
+    fn public_key(&self) -> Option<CoseKey> {
+        None
+    }
+
+    fn sign_1(&self, envelope: CoseSign1) -> Result<CoseSign1, ManyError> {
+        // An anonymous envelope has no signature, or special header.
+        Ok(envelope)
+    }
+}
+
+impl Verifier for AnonymousIdentity {
+    fn sign_1(&self, envelope: &CoseSign1) -> Result<(), ManyError> {
+        if !envelope.signature.is_empty() {
+            Err(ManyError::could_not_verify_signature(
+                "Anonymous should not have a signature.",
+            ))
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// Accept ALL envelopes.
+#[cfg(feature = "testing")]
+mod testing {
+    use crate::Verifier;
+
+    pub struct AcceptAllVerifier;
+
+    impl Verifier for AcceptAllVerifier {
+        fn sign_1(&self, _envelope: &coset::CoseSign1) -> Result<(), many_error::ManyError> {
+            Ok(())
+        }
+    }
+}
+
+#[cfg(feature = "testing")]
+pub use testing::*;
+
+impl Identity for Box<dyn Identity> {
```

Comment:
I don't understand the point of this impl. `AsRef<T>` is implemented for `Box<T>`, so every `Box<dyn T>` can be borrowed as a `&dyn self` ref and call the `&self` methods on it. Doesn't the code compile without these redundant impls? Can you show me it breaking?

Reply:
Nope.

<img width="1181" alt="image" src="https://user-images.githubusercontent.com/681969/185541675-50cb468f-a4f5-42a9-87ff-b78d1a69f923.png">